WordPress/gutenberg
# Add reusable blocks REST API

Baby's first PR! 👶

This adds the REST API for creating, viewing and editing reusable blocks that @westonruter described in https://github.com/WordPress/gutenberg/issues/2081. This API will let us add the ability to turn a static block into a dynamic reusable block (see https://github.com/WordPress/gutenberg/issues/1516).

## What I did

The TL;DR of it all is:

- I'm adding a new post type named `gb_reusable_block`
- I'm adding three REST API endpoints:
    - `GET /wp-json/gutenberg/v1/reusable-blocks` - lists all reusable blocks
    - `GET /wp-json/gutenberg/v1/reusable-blocks/:uuid` - fetches a single reusable block
    - `PUT /wp-json/gutenberg/v1/reusable-blocks/:uuid` - creates or edits a reusable block
- A reusable block looks like this:  
  ```json
  {
    "id": "358b59ee-bab3-4d6f-8445-e8c6971a5605",
    "name": "My cool block",
    "content": "\n<p>Hello <strong>World</strong>! How are you?</p>\n"
  }
  ```
  They're stored as posts with `post_type` set to `gb_reusable_block`.
  - `id` is stored in `post_name`
  - `name` is stored in `post_title`
  - `content` is stored in `post_content`

## How to test

OK. First, make sure you have the [JSON Basic Authentication](https://github.com/WP-API/Basic-Auth/blob/master/basic-auth.php) plugin installed, and that your test WordPress environment has pretty URLs enabled.

Now, you can create a block by hitting the `PUT` endpoint:

```
$ curl -v -X PUT --user admin:password -H 'Content-type: application/json' -d '{"content":"\n<p>Hello!</p>\n","name":"My cool block"}' http://vagrant.local/wp-json/gutenberg/v1/reusable-blocks/358b59ee-bab3-4d6f-8445-e8c6971a5605
```

You'll likely want to change `admin:password` and `http://vagrant.local` to match your test WordPress environment. 

Now that our block is created, we can fetch it:

```
$ curl --user admin:password http://vagrant.local/wp-json/gutenberg/v1/reusable-blocks/358b59ee-bab3-4d6f-8445-e8c6971a5605
{"id":"358b59ee-bab3-4d6f-8445-e8c6971a5605","name":"My cool block","content":"\n<p>Hello!</p>\n"}
```

We can also list all blocks:

```
$ curl --user admin:password http://vagrant.local/wp-json/gutenberg/v1/reusable-blocks
[{"id":"358b59ee-bab3-4d6f-8445-e8c6971a5605","name":"My cool block","content":"\n<p>Hello!</p>\n"}]
```

Play around with creating, editing and reading a few different valid and invalid reusable blocks.

---

cc. @nb @mtias @westonruter
